lib/concurrent/atomic_reference/direct_update.rb

```diff
@@ -27,9 +27,28 @@ def update
     # @!macro [attach] atomic_reference_method_try_update
     #
     # Pass the current value to the given block, replacing it
+    # with the block's result. Return nil if the update fails.
+    #
+    # @yield [Object] Calculate a new value for the atomic reference using
+    #   given (old) value
+    # @yieldparam [Object] old_value the starting value of the atomic reference
+    #
+    # @return [Object] the new value, or nil if update failed
+    def try_update
+      old_value = get
+      new_value = yield old_value
+
+      return unless compare_and_set old_value, new_value
+
+      new_value
+    end
```

Comment:
I think that returning `[success, new_value]` might start adding unneccesary verbosity to the API. When you think about it, `AtomicMarkableReference` would have to then return `[success, [new_mark, new_value]]` which seems needlessly complex. 

Another alternative might be to initialize the value with some sort of private `EmptyObject` or something which is used as a placeholder to distinguish from `nil`.